### PR TITLE
Support for ceilometer-compute-agent on HyperV nodes

### PR DESCRIPTION
**Pull request by aplanas**
Update barclamp to enable the deployment of ceilometer-compute-agent
on hyperv nodes. In the process we split up some recipes and rewrote them.

Several fixes to HyperV installation.

* Update python275.zip from the 2014.2.1 MSI version
* Change nova, neutron and ceilometer to powersell
* Export PBR_VERSION to avoid sdist error during installation
* Remove extra '/dist' during installation
* Update ceilometer 2014.2.1 from official tarballs
* Fix some missing `end` in ceilometer configuration
* Replace `search` call

### CODE REVIEW

**Comment by aplanas** [comment]
Merge together with:
https://github.com/crowbar/barclamp-ceilometer/pull/188
https://github.com/crowbar/barclamp-nova/pull/451

File: chef/cookbooks/hyperv/attributes/default.rb [hunk 1]
```diff
@@ -64,6 +64,15 @@
 default[:openstack][:neutron][:target] = "#{node[:openstack][:location]}"
 default[:openstack][:neutron][:installed] = "#{node[:python][:path]}\\Scripts\\neutron-hyperv-agent.exe"
 
+default[:openstack][:ceilometer][:name] = "ceilometer"
+default[:openstack][:ceilometer][:version] = "2014.2.1"
+default[:openstack][:ceilometer][:file] = "ceilometer-2014.2.1.tar.gz"
+default[:openstack][:leilometer][:target] = "#{node[:openstack][:location]}"
```
Comment (jsuchome) [inline chef/cookbooks/hyperv/attributes/default.rb]: typo ``leilometer``

**Comment by aplanas** [comment]
* Fix `leilometer` typo

File: chef/cookbooks/hyperv/recipes/config_ceilometer.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
+raise if not node[:platform] == 'windows'
```
Comment (jsuchome) [inline chef/cookbooks/hyperv/recipes/config_ceilometer.rb]: What about some error message?

File: chef/cookbooks/hyperv/recipes/config_ceilometer.rb [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  end
+else
+  db_settings = fetch_database_settings
+
+  include_recipe "database::client"
+  include_recipe "#{db_settings[:backend_name]}::client"
+  include_recipe "#{db_settings[:backend_name]}::python-client"
+
+  db_password = ''
+  if node.roles.include? "ceilometer-server"
```
Comment (jsuchome) [inline chef/cookbooks/hyperv/recipes/config_ceilometer.rb]: This should not happen on Windows node, right?

File: chef/cookbooks/hyperv/recipes/config_ceilometer.rb [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+end
+
+
+time_to_live = node[:ceilometer][:database][:time_to_live]
+if time_to_live > 0
+  # We store the value of time to live in days, but config file expects seconds
+  time_to_live = time_to_live * 3600 * 24
+end
+
+is_compute_agent = node.roles.include?("ceilometer-agent") && node.roles.any?{|role| /^nova-multi-compute-/ =~ role}
```
Comment (jsuchome) [inline chef/cookbooks/hyperv/recipes/config_ceilometer.rb]: we should look for ``ceilometer-agent-hyperv`` role here, shouldn't we?

File: chef/cookbooks/hyperv/recipes/config_ceilometer.rb [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    variables(
+      :debug => node[:ceilometer][:debug],
+      :verbose => node[:ceilometer][:verbose],
+      :rabbit_settings => fetch_rabbitmq_settings("ceilometer"),
+      :keystone_settings => keystone_settings,
+      :bind_host => bind_host,
+      :bind_port => bind_port,
+      :metering_secret => node[:ceilometer][:metering_secret],
+      :database_connection => db_connection,
+#      :database_connection => "mongodb://192.168.1.82:27017/ceilometer",
```
Comment (jsuchome) [inline chef/cookbooks/hyperv/recipes/config_ceilometer.rb]: Debugging residue?

File: chef/cookbooks/hyperv/recipes/do_ceilometer.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+raise if not node[:platform] == 'windows'
+
+include_recipe "hyperv::openstack_install_ceilometer"
+include_recipe "hyperv::openstack_install_neutron"
+include_recipe "hyperv::config_neutron"
+include_recipe "hyperv::config_ceilometer"
+include_recipe "hyperv::register_services_ceilometer"
+include_recipe "hyperv::register_services_neutron"
```
Comment (jsuchome) [inline chef/cookbooks/hyperv/recipes/do_ceilometer.rb]: I'm not sure if the order really matters, but usually neutron goes first. Wouldn't it be better to have also here ceilometer parts only after neutron parts?

Comment (jsuchome) [inline chef/cookbooks/hyperv/recipes/do_ceilometer.rb]: One other thought: does it make sense to install ceilometer on hyperv nodes without installing also nova? And if nova is already there, it requires neutron, so ceilometer wouldn't need to require neutron, I think

File: chef/cookbooks/hyperv/recipes/openstack_install_ceilometer.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
+raise if not node[:platform] == 'windows'
+
+#node[:openstack][:git][:clients].each do |openstack_client|
+#  windows_batch "install_openstack_clients" do
+#    code <<-EOH
+#    #{node[:pip][:command]} install --upgrade #{openstack_client}
+#    EOH
+#   # not_if {::File.exists?(node[:pymysql][:installed])}
+#  end
+#end
```
Comment (jsuchome) [inline chef/cookbooks/hyperv/recipes/openstack_install_ceilometer.rb]: Is there any reason to keep commented parts?

File: chef/cookbooks/hyperv/recipes/openstack_install_nova.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
+raise if not node[:platform] == 'windows'
+
+#node[:openstack][:git][:clients].each do |openstack_client|
+#  windows_batch "install_openstack_clients" do
+#    code <<-EOH
+#    #{node[:pip][:command]} install --upgrade #{openstack_client}
+#    EOH
+#   # not_if {::File.exists?(node[:pymysql][:installed])}
+#  end
+#end
```
Comment (jsuchome) [inline chef/cookbooks/hyperv/recipes/openstack_install_nova.rb]: Same as above

File: chef/cookbooks/hyperv/recipes/do_ceilometer.rb (hunk 6 above)
Comment (aplanas) [inline chef/cookbooks/hyperv/recipes/do_ceilometer.rb]: Technically makes sense to install Ceilometer agent w/out Nova.  AFAIK Ceilometer can read from Nova, Cinder, Swift, Neutron ... and others.  But I agree that the normal use-case is Nova here.

Comment (jsuchome) [inline chef/cookbooks/hyperv/recipes/do_ceilometer.rb]: In general, of course, but we're talking about hyperv nodes only, and there couldn't be any other services than nova and ceilometer, AFAIK.

Comment (aplanas) [inline chef/cookbooks/hyperv/recipes/do_ceilometer.rb]: True. But there are some bits that IMHO makes this approach better:

* The `role_constrains` is decoupled.  If we allows some new agent to be installed in a Windows machine, we need to rewrite this part to separate the code as is done now.
* The barclamp-ceilometer allows you to install the agent independently from what is installed in the node.  This is true for GNU/Linux too
* The barclamp-ceilometer suggest automatically the correct Hyper-V node when you want to deploy Ceilometer in the cloud.

So, basically, if we constrain the installation of Ceilometer in HyperV nodes only when you have deployed Nova, we behave differently for the rest of the barclamp-ceilometer.

File: chef/cookbooks/hyperv/recipes/config_ceilometer.rb (hunk 2 above)
Comment (nkrinner) [inline chef/cookbooks/hyperv/recipes/config_ceilometer.rb]: In the past we did not have any error messages here, but we could add them of course.

File: chef/cookbooks/hyperv/recipes/do_ceilometer.rb (hunk 6 above)
Comment (jsuchome) [inline chef/cookbooks/hyperv/recipes/do_ceilometer.rb]: Hm, right, OK. I think it might be good to change barclamp-ceilometer to include such dependency, but let's keep it somple now.

File: chef/cookbooks/hyperv/recipes/config_ceilometer.rb (hunk 4 above)
Comment (nkrinner) [inline chef/cookbooks/hyperv/recipes/config_ceilometer.rb]: It is correct that we look for the wrong ceilometer-agent role here. I suggest to implement a match query as it is done with the nova compute node part  and also move that to the original ceilometer barclamp.

File: chef/cookbooks/hyperv/recipes/config_ceilometer.rb (hunk 5 above)
Comment (nkrinner) [inline chef/cookbooks/hyperv/recipes/config_ceilometer.rb]: Yes. Needs to be dropped.

File: chef/cookbooks/hyperv/recipes/do_ceilometer.rb (hunk 6 above)
Comment (nkrinner) [inline chef/cookbooks/hyperv/recipes/do_ceilometer.rb]: I think switching the order in which the services are registered makes sense.

**Comment by nkrinner** [comment]
There are some parts that are not needed that are copied from the original ceilometer barclamp. I personally prefer to keep them here / to unify both barclamps to have the code easier to read and understand and also to allow easier maintenance via copy and paste in the future.

File: chef/cookbooks/hyperv/attributes/default.rb [hunk 9]
```diff
@@ -64,6 +64,15 @@
 default[:openstack][:neutron][:target] = "#{node[:openstack][:location]}"
 default[:openstack][:neutron][:installed] = "#{node[:python][:path]}\\Scripts\\neutron-hyperv-agent.exe"
 
+default[:openstack][:ceilometer][:name] = "ceilometer"
+default[:openstack][:ceilometer][:version] = "2014.2.1"
+default[:openstack][:ceilometer][:file] = "ceilometer-2014.2.1.tar.gz"
+default[:openstack][:ceilometer][:target] = "#{node[:openstack][:location]}"
+default[:openstack][:ceilometer][:installed] = "#{node[:python][:path]}\\Scripts\\ceilometer-agent-compute.exe"
+default[:openstack][:ceilometer][:lock_path] = "C:\\OpenStack\\var\\run\\"
+default[:openstack][:ceilometer][:log_dir] = "C:\\OpenStack\\log\\"
```
Comment (vuntz) [inline chef/cookbooks/hyperv/attributes/default.rb]: Any reason to not simply re-use `default[:openstack][:log]`?

File: chef/cookbooks/hyperv/recipes/config_neutron.rb [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    action :create
+    recursive true
+  end
+end
+
+%w{ OpenStackService.exe mkisofs.exe mkisofs_license.txt qemu-img.exe intl.dll libglib-2.0-0.dll libssp-0.dll zlib1.dll }.each do |bin_file|
+  cookbook_file "#{node[:openstack][:bin]}/#{bin_file}" do
+    source bin_file
+  end
+end
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/config_neutron.rb]: Why is this in the neutron recipe? At least some of this is only for nova.

File: chef/cookbooks/hyperv/recipes/config_neutron.rb [hunk 11]
```diff
@@ -0,0 +1,64 @@
+raise if not node[:platform] == 'windows'
+
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/config_neutron.rb]: Shouldn't all the code here be dropped from `config_nova.rb`?

File: chef/cookbooks/hyperv/recipes/do_ceilometer.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+raise if not node[:platform] == 'windows'
+
+include_recipe "hyperv::openstack_install_ceilometer"
+include_recipe "hyperv::openstack_install_neutron"
+include_recipe "hyperv::config_neutron"
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/do_ceilometer.rb]: Why neutron here? Ceilometer doesn't depend on it. The only reason we install it is for nova.

**Comment by aplanas** [comment]
* Remove code for 'ceilometer-server' role
* Fix roles references in `is_compute_agent`
* Remove dead / commented code
* Remove Neutron installation from 'do_ceilometer.rb'

**Comment by aplanas** [comment]
* Remove more commented code
* Reuse default log dir in ceilometer recipe

File: chef/cookbooks/hyperv/recipes/config_neutron.rb (hunk 10 above)
Comment (aplanas) [inline chef/cookbooks/hyperv/recipes/config_neutron.rb]: @nkrinner @ovisan, I guess that OpenStackService.exe needs to be there for Neutron, but are "intl.dll", "libglib-2.0-0.dll", "libssp-0.dll", "zlib1.dll" dependencies for qemu, or for OpenStackService.exe?

File: chef/cookbooks/hyperv/recipes/config_neutron.rb (hunk 11 above)
Comment (aplanas) [inline chef/cookbooks/hyperv/recipes/config_neutron.rb]: Dropped from `confirg_nova.rb`

File: chef/cookbooks/hyperv/recipes/config_neutron.rb (hunk 10 above)
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/config_neutron.rb]: I think they're dependencies of the qemu-img.exe binary.

File: chef/cookbooks/hyperv/recipes/config_nova.rb [hunk 13]
```diff
@@ -88,25 +88,3 @@
...
-vlan_end = [vlan_start + num_vlans - 1, 4094].min
-
-template "#{node[:openstack][:config].gsub(/\\/, "/")}/neutron_hyperv_agent.conf" do
-  source "neutron_hyperv_agent.conf.erb"
-  variables(
-            :rabbit_settings => fetch_rabbitmq_settings("nova"),
-            :openstack_location => node[:openstack][:location],
-            :openstack_log => node[:openstack][:log],
-            :neutron_networking_plugin => neutron_networking_plugin,
-            :neutron_ml2_type_drivers => neutron_ml2_type_drivers,
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/config_nova.rb]: I think you can drop the definition of this variable from earlier in the recipe.

Comment (aplanas) [inline chef/cookbooks/hyperv/recipes/config_nova.rb]: True.  I am also for removing glance and cinder references.

**Comment by aplanas** [comment]
* Remove references to `neutron_ml2_type_drivers` in `config_nova.rb`
* Remove unused binary files from `config_neutron.rb`

File: chef/cookbooks/hyperv/recipes/config_neutron.rb [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    action :create
+    recursive true
+  end
+end
+
+%w{ OpenStackService.exe }.each do |bin_file|
+  cookbook_file "#{node[:openstack][:bin]}/#{bin_file}" do
+    source bin_file
+  end
+end
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/config_neutron.rb]: So the creation of the directories and of this exec is something that is done (or if not, should be done) in the three config recipes. Would it make sense to have a config_common_openstack.rb recipe for that kind of things?

**Comment by aplanas** [comment]
* Extract directory creation and OpenStackService.exe installation in `config_common_openstack.rb`
* Call `config_common_openstack.rb` recipe from `do_ceilometer.rb` and `do_nova.rb`
* Reorder `do_nova.rb` to deploy and configure neutron first

**Comment by aplanas** [comment]
@ovisan Can you help me testing this in c1 cloud?

**Comment by ovisan** [comment]
@aplanas  sure thing

**Comment by jsuchome** [comment]
I like it.... +1 assuming tests confirm it works

File: chef/cookbooks/hyperv/recipes/config_ceilometer.rb [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+end
+
+
+time_to_live = node[:ceilometer][:database][:time_to_live]
+if time_to_live > 0
+  # We store the value of time to live in days, but config file expects seconds
+  time_to_live = time_to_live * 3600 * 24
+end
+
+is_compute_agent = %w(ceilometer-agent-hyperv nova-multi-compute-hyperv).all?{|role| roles.include?(role)}
```
Comment (nkrinner) [inline chef/cookbooks/hyperv/recipes/config_ceilometer.rb]: I think it has to be `node.roles.include?(role)` , not `roles.include?(role)`

**Comment by aplanas** [comment]
* Fix reference to `node.roles.include?(role)`

**Comment by ovisan** [comment]
Still needs refactoring
```←[0m
================================================================================←[0m
←[31mRecipe Compile Error in C:/chef/cache/cookbooks/hyperv/recipes/do_ceilometer.rb←[0m
================================================================================←[0m

←[0m
NameError←[0m
---------←[0m
Cannot find a resource for roles on windows version 6.3.9600←[0m

←[0m
Cookbook Trace:←[0m
---------------←[0m
  C:/chef/cache/cookbooks/hyperv/recipes/config_ceilometer.rb:57:in `block in from_file'
  C:/chef/cache/cookbooks/hyperv/recipes/config_ceilometer.rb:57:in `each'
  C:/chef/cache/cookbooks/hyperv/recipes/config_ceilometer.rb:57:in `all?'
  C:/chef/cache/cookbooks/hyperv/recipes/config_ceilometer.rb:57:in `from_file'
  C:/chef/cache/cookbooks/hyperv/recipes/do_ceilometer.rb:5:in `from_file'←[0m

←[0m
Relevant File Content:←[0m
----------------------←[0m
C:/chef/cache/cookbooks/hyperv/recipes/config_ceilometer.rb:

 50:
 51:  time_to_live = node[:ceilometer][:database][:time_to_live]
 52:  if time_to_live > 0
 53:    # We store the value of time to live in days, but config file expects seconds
 54:    time_to_live = time_to_live * 3600 * 24
 55:  end
 56:
 57>> is_compute_agent = %w(ceilometer-agent-hyperv nova-multi-compute-hyperv).all?{|role| roles.include?(role)}
 58:
 59:  dirs = [ node[:openstack][:ceilometer][:lock_path], node[:openstack][:log], node[:openstack][:ceilometer][:signing
_dir] ]
 60:  dirs.each do |dir|
 61:    directory dir do
 62:      action :create
 63:      recursive true
 64:    end
 65:  end
 66:
←[0m

←[0m
Chef Client failed. 0 resources updated←[0m```
https://guides.github.com/features/mastering-markdown/

**Comment by aplanas** [comment]
@ovisan Can you redeploy the last version? This was fixed in the last commit.

**Comment by ovisan** [comment]
@aplanas i did fetch the new code and installed the barclamp, this is the latest code.

**Comment by aplanas** [comment]
@ovisan I am sure that this test was not running the latest code. There is a reference to `{|role| roles.include?(role)}` in the error message, that is the one that @nkrinner pointed in https://github.com/crowbar/barclamp-hyperv-compute/pull/32#issuecomment-101250318 that is fixed in https://github.com/aplanas/barclamp-hyperv-compute/commit/f1601ab22e3b211f2e48b9b2f84c48b2c735ed87#diff-160729b42a011a7738a612f4c9c16e2dR57

**Comment by nkrinner** [comment]
@ovisan Line 57 in config_ceilometer.rb is now different than what you deployed. Maybe there was an error when updating the cookbook on the server?

**Comment by ovisan** [comment]
there is confusion, the ceilometer barclamp fails, not the hyperv one. Should have commented on that one

**Comment by nkrinner** [comment]
@ovisan the ceilometer cookbook calls the recipes hyperv cookbook, where the error happens. So although the error is triggered when ceilometer is deployed, the faulty code lives in the hyperv barclamp (afaics).

**Comment by ovisan** [comment]
Pulled the latest changes and tested again:

```
================================================================================←[0m
←[31mRecipe Compile Error in C:/chef/cache/cookbooks/hyperv/recipes/do_ceilometer.rb←[0m
================================================================================←[0m

←[0m
NoMethodError←[0m
-------------←[0m
undefined method `[]' for nil:NilClass←[0m

←[0m
Cookbook Trace:←[0m
---------------←[0m
  C:/chef/cache/cookbooks/hyperv/recipes/openstack_install_ceilometer.rb:3:in `from_file'
  C:/chef/cache/cookbooks/hyperv/recipes/do_ceilometer.rb:3:in `from_file'←[0m

←[0m
Relevant File Content:←[0m
----------------------←[0m
C:/chef/cache/cookbooks/hyperv/recipes/openstack_install_ceilometer.rb:

  1:  raise if not node[:platform] == 'windows'
  2:
  3>> cookbook_file "#{node[:cache_location]}#{node[:openstack][:ceilometer][:file]}" do
  4:    source node[:openstack][:ceilometer][:file]
  5:    not_if {::File.exists?(node[:openstack][:ceilometer][:installed])}
  6:  end
  7:
  8:  windows_batch "unzip_ceilometer" do
  9:    code <<-EOH
 10:    #{node[:sevenzip][:command]} x #{node[:cache_location]}#{node[:openstack][:ceilometer][:file]} -o#{node[:opensta
ck][:location]} -r -y
 11:    #{node[:sevenzip][:command]} x #{node[:openstack][:location]}\\dist\\#{node[:openstack][:ceilometer][:name]}-#{n
ode[:openstack][:ceilometer][:version]}.tar -o#{node[:openstack][:location]} -r -y
 12:    rmdir /S /Q #{node[:openstack][:location]}\\dist
←[0m

←[0m
Chef Client failed. 0 resources updated←[0m
```

**Comment by aplanas** [comment]
@nkrinner any clue here? The complains looks like that is in `node[:openstack][:ceilometer][:file]`.  `node[:openstack]` or `https://trello.com/c/uAQi0rn6/872-training-upgrading-suse-openstack-cloud` is `nil`

**Comment by ovisan** [comment]
and like https://bugzilla.suse.com/show_bug.cgi?id=928714

**Comment by aplanas** [comment]
@ovisan me and @nkrinner try to reproduce the bug.  We were unable to replicate it and the deployment was OK.

**Comment by aplanas** [comment]
reviews?

**Comment by ovisan** [comment]
@aplanas which cloud? are all 3 services running?

**Comment by ovisan** [comment]
:+1:

File: chef/cookbooks/hyperv/recipes/config_ceilometer.rb [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+
+time_to_live = node[:ceilometer][:database][:time_to_live]
+if time_to_live > 0
+  # We store the value of time to live in days, but config file expects seconds
+  time_to_live = time_to_live * 3600 * 24
+end
+
+is_compute_agent = %w(ceilometer-agent-hyperv nova-multi-compute-hyperv).all?{|role| node.roles.include?(role)}
+
+dirs = [ node[:openstack][:ceilometer][:lock_path], node[:openstack][:log], node[:openstack][:ceilometer][:signing_dir] ]
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/config_ceilometer.rb]: `node[:openstack][:log]` can be dropped here (done in the common recipe)

File: chef/cookbooks/hyperv/recipes/config_neutron.rb [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+  neutron_server_port = nil
+  neutron_service_user = nil
+  neutron_service_password = nil
+  neutron_networking_plugin = "ml2"
+  neutron_ml2_type_drivers = []
+end
+Chef::Log.info("Neutron server at #{neutron_server_host}")
+
+# Chef 11.4 fails to notify if the path separator is windows like, according to https://tickets.opscode.com/browse/CHEF-4082
+# using gsub to replace the windows path separator to linux one
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/config_neutron.rb]: that comment should likely just go away, or be just before all templates where we do `gsub(/\\/, "/")`

File: chef/cookbooks/hyperv/recipes/register_services_ceilometer.rb [hunk 18]
```diff
@@ -0,0 +1,21 @@
+raise if not node[:platform] == 'windows'
+
+powershell "register_services" do
+  code <<-EOH
+    if (-not (Get-Service "#{node[:service][:ceilometer][:name]}" -ErrorAction SilentlyContinue))
+    {
+      # $secpasswd = ConvertTo-SecureString "crowbar" -AsPlainText -Force
+      # $credentials = New-Object System.Management.Automation.PSCredential (".\\Administrator", $secpasswd)
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/register_services_ceilometer.rb]: Any reason to keep these comments? Especially as we have the password configurable now, which make them wrong.

**Comment by vuntz** [comment]
Minor nitpicks; feel free to consider my comment as +1 once they're fixed.

**Comment by aplanas** [comment]
* Remove re-creation of log directory
* Move comment into the proper place
* Remove some debug comments

File: chef/cookbooks/hyperv/recipes/register_services_nova.rb [hunk 19]
```diff
@@ -0,0 +1,21 @@
+raise if not node[:platform] == 'windows'
+
+powershell "register_services" do
+  code <<-EOH
+    if (-not (Get-Service "#{node[:service][:nova][:name]}" -ErrorAction SilentlyContinue))
+    {
+      # $secpasswd = ConvertTo-SecureString "crowbar" -AsPlainText -Force
+      # $credentials = New-Object System.Management.Automation.PSCredential (".\\Administrator", $secpasswd)
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/register_services_nova.rb]: Heh, sorry: forgot to mention that the comments are also here (as well as in the neutron recipe ;-))

File: chef/cookbooks/hyperv/recipes/config_ceilometer.rb [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+      :lock_path => node[:openstack][:ceilometer][:lock_path],
+      :log_dir =>  node[:openstack][:log],
+      :signing_dir => node[:openstack][:ceilometer][:signing_dir]
+    )
+    if is_compute_agent
+      notifies :restart, "service[nova-compute]"
+    end
+end
+
+template "#{node[:openstack][:config].gsub(/\\/, "/")}/pipeline.yaml" do
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/config_ceilometer.rb]: No `# Chef 11.4 fails to notify if the path separator is windows like, ...` comment here?

**Comment by aplanas** [comment]
* Remove more debuging comments from register_service_*
* Normalize comments in config_*

**Comment by vuntz** [comment]
+1

**Comment by aplanas** [comment]
@nkrinner @ovisan reviews?
